datafusion/src/physical_plan/expressions/bitmap_distinct.rs

```diff
@@ -0,0 +1,233 @@
+                    &array
+                        .values()
+                        .iter()
+                        .map(|&x| x as u32)
+                        .collect::<Vec<u32>>(),
+                );
+            }
+            DataType::Int16 => {
+                let array = value.as_any().downcast_ref::<Int16Array>().unwrap();
+                self.bitmap.add_many(
+                    &array
+                        .values()
+                        .iter()
+                        .map(|&x| x as u32)
+                        .collect::<Vec<u32>>(),
+                );
+            }
+            DataType::Int32 => {
+                let array = value.as_any().downcast_ref::<Int32Array>().unwrap();
+                self.bitmap.add_many(
+                    &array
+                        .values()
+                        .iter()
+                        .map(|&x| x as u32)
+                        .collect::<Vec<u32>>(),
+                );
+            }
+            DataType::UInt8 => {
+                let array = value.as_any().downcast_ref::<UInt8Array>().unwrap();
+                self.bitmap.add_many(
+                    &array
+                        .values()
+                        .iter()
+                        .map(|&x| x as u32)
+                        .collect::<Vec<u32>>(),
+                );
+            }
+            DataType::UInt16 => {
+                let array = value.as_any().downcast_ref::<UInt16Array>().unwrap();
+                self.bitmap.add_many(
+                    &array
+                        .values()
+                        .iter()
+                        .map(|&x| x as u32)
+                        .collect::<Vec<u32>>(),
+                );
+            }
+            DataType::UInt32 => {
+                let array = value.as_any().downcast_ref::<UInt32Array>().unwrap();
+                self.bitmap.add_many(array.values());
+            }
+            e => {
+                return Err(DataFusionError::Internal(format!(
+                    "BITMAP_COUNT_DISTINCT is not expected to receive the type {:?}",
+                    e
+                )));
+            }
+        }
+        Ok(())
+    }
+
+    fn merge_batch(&mut self, states: &[ArrayRef]) -> Result<()> {
+        assert_eq!(1, states.len(), "expect only 1 element in the states");
+        let binary_array = states[0].as_any().downcast_ref::<BinaryArray>().unwrap();
+        let bitmaps = binary_array
+            .iter()
+            .map(|x| {
+                Bitmap::deserialize(x.expect(
+                    "Impossibly got empty binary array from states in bitmap_distinct",
+                ))
+            })
+            .collect::<Vec<Bitmap>>();
+        let bitmaps = &bitmaps.iter().map(|x| x.borrow()).collect::<Vec<&Bitmap>>()[..];
+        //Do not use self.bitmap = xxx, because '=' has been wrote for 'BitAnd' !.
+        self.bitmap.bitor_assign(Bitmap::fast_or(bitmaps));
+        Ok(())
```

Comment:
I might be way off here (sorry if so), but when I look at this I wonder if you can simplify it to this:

```suggestion
        for data in binary_array.iter().flatten() {
            self.bitmap.or_inplace(&Bitmap::deserialize(data));
        }
        Ok(())
```

Reply:
Thanks, I used to use `or_inplace `
copy from [cRoaring](https://github.com/RoaringBitmap/CRoaring#dealing-with-large-volumes)
> Some users have to deal with large volumes of data. It may be important for these users to be aware of the addMany (C++) roaring_bitmap_or_many (C) functions as it is much faster and economical to add values in batches when possible. Furthermore, calling periodically the runOptimize (C++) or roaring_bitmap_run_optimize (C) functions may help.

`fast_or ` bottom use `roaring_bitmap_or_many`,  but i tested  local, there are little difference between them, maybe i will test in large volumes data.